### PR TITLE
Fix CA2211 warnings

### PR DESCRIPTION
**Pull request by kartheekp-ms**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/648

Regression? Last working version: N/A

## Description

Fix [CA2211](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2211) warnings in the product code. In couple of cases, added justification to the existing suppression why I think the warning can't be fixed.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] N/A 

- **Documentation**
   
  - [x] N/A

### CODE REVIEW

**Review by erdembayar — APPROVED** [review]
👍

File: src/NuGet.Clients/NuGet.Console/GlobalSuppressions.cs [hunk 1]
```diff
@@ -13,8 +13,7 @@
...
 [assembly: SuppressMessage("Usage", "VSTHRD002:Avoid problematic synchronous waits", Justification = "Dispose method", Scope = "member", Target = "~M:NuGetConsole.ChannelOutputConsole.Dispose(System.Boolean)")]
-[assembly: SuppressMessage("Build", "CA2211:Non-constant fields should not be visible", Justification = "<Pending>", Scope = "member", Target = "~F:NuGetConsole.GuidList.guidNuGetOutputWindowPaneGuid")]
-[assembly: SuppressMessage("Build", "CA2211:Non-constant fields should not be visible", Justification = "<Pending>", Scope = "member", Target = "~F:NuGetConsole.GuidList.guidVsWindowKindOutput")]
+[assembly: SuppressMessage("Build", "CA2211:Non-constant fields should not be visible", Justification = "Cannot change access modifier of 'guidNuGetOutputWindowPaneGuid' because it is referred in another assembly. This variable is passed as ref paramter to VS API and hence it can't be const/readonly.", Scope = "member", Target = "~F:NuGetConsole.GuidList.guidNuGetOutputWindowPaneGuid")]
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.Console/GlobalSuppressions.cs]: This project isn't part of the NuGet SDK. Changing the field to a property should make the analyzer happy as well.

File: src/NuGet.Core/NuGet.Packaging.Extraction/GlobalSuppressions.cs [hunk 2]
```diff
@@ -7,7 +7,7 @@
 
 [assembly: SuppressMessage("Design", "CA1056:Uri properties should not be strings", Justification = "<Pending>", Scope = "member", Target = "~P:NuGet.Packaging.Core.RepositoryMetadata.Url")]
 [assembly: SuppressMessage("Design", "CA1012:Abstract types should not have constructors", Justification = "<Pending>", Scope = "type", Target = "~T:NuGet.Packaging.PackageReaderBase")]
-[assembly: SuppressMessage("Build", "CA2211:Non-constant fields should not be visible", Justification = "<Pending>", Scope = "member", Target = "~F:NuGet.Packaging.Licenses.NuGetLicenseData.LicenseListVersion")]
+[assembly: SuppressMessage("Build", "CA2211:Non-constant fields should not be visible", Justification = "Code is auto-generated by the GenerateLicenseList tool.", Scope = "member", Target = "~F:NuGet.Packaging.Licenses.NuGetLicenseData.LicenseListVersion")]
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging.Extraction/GlobalSuppressions.cs]: The GenerateLicenseList tool is our own custom tool, and I believe lives in the same repo, so we could change it. However, it's part of the public API, which is a stronger reason why we possibly shouldn't change it. On the other hand, you introduced API breaking changes in NuGet.ProjectModel and NuGet.Versioning, so it seems like this field should be consistent.

File: src/NuGet.Clients/NuGet.Console/GlobalSuppressions.cs (hunk 1 above)
Comment (kartheekp-ms) [reply src/NuGet.Clients/NuGet.Console/GlobalSuppressions.cs]: When I changed `field` to `property` -> `public static Guid guidNuGetOutputWindowPaneGuid => Guid.Parse("CEC55EC8-CC51-40E7-9243-57B87A6F6BEB"); ` compiler raises an error in another assembly.

image

Comment (kartheekp-ms) [reply src/NuGet.Clients/NuGet.Console/GlobalSuppressions.cs]: I have updated `justification` in the latest commit to be clearer.

File: src/NuGet.Core/NuGet.Packaging.Extraction/GlobalSuppressions.cs (hunk 2 above)
Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Packaging.Extraction/GlobalSuppressions.cs]: Fixed in the latest commit. updated `$\NuGet.Client\test\TestExtensions\GenerateLicenseList\LicenseDataCodeGenerator.cs` file so that this change won't be reverted in the future.

File: src/NuGet.Clients/NuGet.Console/GlobalSuppressions.cs [hunk 3]
```diff
@@ -9,12 +9,9 @@
 [assembly: SuppressMessage("Microsoft.Design", "CA1020:AvoidNamespacesWithFewTypes", Scope = "namespace", Target = "NuGetConsole.Implementation")]
 [assembly: SuppressMessage("Microsoft.Design", "CA1020:AvoidNamespacesWithFewTypes", Scope = "namespace", Target = "NuGetConsole.Implementation.Console")]
 [assembly: SuppressMessage("Microsoft.Design", "CA1020:AvoidNamespacesWithFewTypes", Scope = "namespace", Target = "NuGetConsole.Implementation.PowerConsole")]
-[assembly: SuppressMessage("Microsoft.Design", "CA1020:AvoidNamespacesWithFewTypes", Scope = "namespace", Target = "NuGetConsole.DebugConsole")]
-[assembly: SuppressMessage("Microsoft.Design", "CA1020:AvoidNamespacesWithFewTypes", Scope = "namespace", Target = "NuGetConsole.Host.PowerShell")]
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.Console/GlobalSuppressions.cs:13]: IDE complained as unnecessary suppressions. I couldn't find the `Target` types in the product code.

File: src/NuGet.Clients/NuGet.Console/GlobalSuppressions.cs (hunk 1 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.Console/GlobalSuppressions.cs]: You can use a temporary variable, so the code doesn't need to try to use ref on a property.
```cs
Guid tempGuid = Guid.Zero;
if (_vsOutputWindow.Value.GetPane(ref tempGuid, out pane) == VSConstances.S_OK)
{
  NuGetConsole.Guidlist.guidNuGetOuputWindowPaneGuid = tempGuid;
  pane.Activate();
}
```

Although the name should probably be better than `tempGuid`.

Comment (kartheekp-ms) [reply src/NuGet.Clients/NuGet.Console/GlobalSuppressions.cs]: > When I changed field to property -> public static Guid guidNuGetOutputWindowPaneGuid => Guid.Parse("CEC55EC8-CC51-40E7-9243-57B87A6F6BEB"); compiler raises an error in another assembly.

My above comment is incorrect. `public static guidNuGetOutputWindowPaneGuid` property should have both `getter` and `setter` instead of `getter` alone because of usage in another assembly. I am thinking of what would be the benefit of replacing `public static` `field` with a `property` other than making the analyzer happy. There are three instances in the product code where we have similar code that passes `guidNuGetOutputWindowPaneGuid` as `ref` parameter. I could be wrong, but I lean towards keeping the suppression with justification as is. Happy to learn more if my analysis is incorrect.

image

Comment (zivkan) [reply src/NuGet.Clients/NuGet.Console/GlobalSuppressions.cs]: >  I am thinking of what would be the benefit of replacing public static field with a property other than making the analyzer happy

This is the general convention of .NET libraries. Some people believe that fields should never, ever be public, only used privately in classes. Only make methods, properties and const available in public APIs.

The reason is that fields do not provide any way to do input validation on setting. Whereas with a property setter you can. Having a auto property that you later change to a manual property with validation is not an API breaking change, whereas changing a field to a property is.

Comment (zivkan) [reply src/NuGet.Clients/NuGet.Console/GlobalSuppressions.cs]: I believe this .NET convention of avoiding public fields is a carry over from Java Enterprise development, where properties don't exist, and public fields are strongly discouraged in favor or methods starting with `get` and `set`, since they can do the appropriate input validation.

File: src/NuGet.Clients/NuGet.CommandLine/CommandLineParser.cs [hunk 4]
```diff
@@ -18,7 +18,6 @@ public class CommandLineParser
         private readonly ICommandManager _commandManager;
 
         // On Unix or MacOSX slash as a switch indicator would interfere with the path separator
-        [SuppressMessage("Microsoft.Performance", "CA1802:UseLiteralsWhereAppropriate")]
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.CommandLine/CommandLineParser.cs:21]: IDE complained as unnecessary suppression.

File: src/NuGet.Clients/NuGet.Console/GlobalSuppressions.cs (hunk 1 above)
Comment (kartheekp-ms) [reply src/NuGet.Clients/NuGet.Console/GlobalSuppressions.cs]: Thanks for the feedback Andy. Since there is no validation during `setter` I was unsure of the benefit. But I liked your suggestion because changing field to property is a breaking change. If we change the `field` to `property` now we can add validation logic to `setter` in the future without a breaking change.

Addressed this comment in the latest commit. Please review.

**Review by erdembayar — APPROVED** [review]

**Comment by kartheekp-ms** [comment]
Waiting for second approval or additional feedback.

File: src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseData.cs [hunk 5]
```diff
@@ -13,7 +13,7 @@ public static class NuGetLicenseData
         // Refer to https://github.com/spdx/LicenseListPublisher/issues/102. 
         // When this files gets regenerated, this explanation will be removed. 
         // This is to be used as a marker to confirm that the linked issue with the data source has been fixed.
-        public static string LicenseListVersion = "da7ecca";
+        public const string LicenseListVersion = "da7ecca";
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseData.cs]: This value is not a constant. It changes each time the SPDX owners update their data source, and then we run our tool to regenerate this file.

Making it a const will mean that code that reference this value, but then have a different version of `NuGet.Packaging` brought in because of dependencies, will have mismatched `LicenseListVersion` compared to the `LisenseList` below.

Making `LicenseListVersion` a get-only property would be better.

Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseData.cs]: AFAIK, we don't need to change `LicenseListVersion` to `readonly` as per my understanding.
- To update `NuGetLicenseData` class file we execute [`$repositoryRoot/scripts/utils/UpdateNuGetLicenseSPDXList.ps1`](https://github.com/NuGet/NuGet.Client/blob/dev/scripts/utils/UpdateNuGetLicenseSPDXList.ps1)
- `$targetFile` parameter in the above script is set to `src\NuGet.Core\NuGet.Packaging\Licenses\NuGetLicenseData.cs`
- The contents of the `$targetFile` are overwritten [here](https://github.com/NuGet/NuGet.Client/blob/dev/test/TestExtensions/GenerateLicenseList/Program.cs#L38). The [File.CreateText(String) Method doc](https://docs.microsoft.com/en-us/dotnet/api/system.io.file.createtext?view=net-5.0) says `If the file already exists, its contents are overwritten`.
- I modified the logic that generates the class file in this PR to always mark the variable as `const` https://github.com/NuGet/NuGet.Client/pull/4028/files#diff-26391a81ca1c15b62f62a482e277541b458d7edd4013b2f12e4fefb82caedb9cR123

Please let me know if I am missing something. I am happy to learn more and change the code accordingly.

Comment (zivkan) [reply src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseData.cs]: How the file or value gets updated doesn't matter to customers using our package. It's just our own implementation detail, which we can change.

From the customer point of view:

|If `LicenseListVersion` is|then|
|---|---|
|a writable field|The compiler won't protect them, or any other library they use, from accidentally setting a different value.|
|A const field|The value they compiled against gets "burned in" to their assembly, and if at runtime they actually use a different version of NuGet.Packaging with a different version of the SPDX data, then the version that's burned in to their assembly won't match the version burned in our assembly.|
|A get only property|The version of the SPDX data used in NuGet.Packaging will always be retrieved, even if a different version of NuGet.Packaging is used at runtime, compared to at compile time.|
|A readonly field|Like the read-only property it will always get the right version at runtime, not compile time. But it prevents refactoring of the public API where the implementation might not want to be a field, or might want to be stored somewhere else. A property is just a fancy method, and the method/property getter can be modified during a refactoring to maintain API and ABI compatibility, whereas to maintain API and ABI compatibility with a field means the field can never change. In tight loops, a field will have better perf than a method, but that's not a concern here.|


In short, change `LicenseListVersion` to `public static string LicenseListVersion { get; } = "abc123";` or `public static string LicenseListVersion { get => "abc123" }`

Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseData.cs]: Thank you Andy, for the detailed explanation. I changed the field to a `get only property`.

**Comment by kartheekp-ms** [comment]
I have addressed all the comments on this PR. Waiting for second approval or additional feedback. Planning to merge in the next 2 to 3 days.

**Comment by kartheekp-ms** [comment]
@NuGet/nuget-client 🔔  waiting for second approval.

**Comment by kartheekp-ms** [comment]
@NuGet/nuget-client 🔔 waiting for second approval. Addressed all the comments.

**Review by nkolev92 — APPROVED** [review]
Fine by me.

I didn't intend to review this as others have already looked into it, but it's been a while :)
